### PR TITLE
feat(cairo_native): integrate native compilation into CommandLineCompiler

### PR DESCRIPTION
**Pull request by avi-starkware**

### CODE REVIEW

**Review by ArniStarkware — CHANGES_REQUESTED** [review]
> Reviewed 2 of 10 files at r1.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1998)* status: 2 of 10 files reviewed, 2 unresolved discussions (waiting on @avi-starkware and @Yoni-Starkware)

___
*[`crates/starknet_sierra_compile/src/command_line_compiler.rs` line 28 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/1998#-OBZolRN4mvlNkNUMiDe:-OBZolRN4mvlNkNUMiDf:bn4mvqh) ([raw file](https://github.com/starkware-libs/sequencer/blob/6b16bfb4d9f32eef82023b09eaef0dd1855d59dd/crates/starknet_sierra_compile/src/command_line_compiler.rs#L28)):*
> ```XML
>     #[cfg(feature = "cairo_native")]
>     path_to_starknet_native_compile_binary: PathBuf,
> }
> ```

Will this structure typically be used for the compilations `Sierra` -> `Casm` *and* `Sierra` -> `Cairo-native`?
Will there be a case where we want to have the second feature without the first? In that case, setting up the first feature is redundant. Maybe clever use of traits can save code duplication here.



_Suggestion:_
```XML
#[derive(Clone)]
pub struct CommandLineCompiler {
    pub config: SierraToCasmCompilationConfig, // TODO: use the correct config 0 that is relevant for any config.
    path_to_compiler_binary: PathBuf,
}
```
___
*[`crates/starknet_sierra_compile/src/build_utils.rs` line 5 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/1998#-OBZo-aT5-7zks4SDF5A:-OBZo-aT5-7zks4SDF5B:b-k1shsn) ([raw file](https://github.com/starkware-libs/sequencer/blob/6b16bfb4d9f32eef82023b09eaef0dd1855d59dd/crates/starknet_sierra_compile/src/build_utils.rs#L5)):*
> ```XML
> pub(crate) const CAIRO_LANG_BINARY_NAME: &str = "starknet-sierra-compile";
> #[cfg(feature = "cairo_native")]
> pub(crate) const NATIVE_BINARY_NAME: &str = "starknet-native-compile";
> ```

Or is it redundant?



_Suggestion:_
```XML
pub(crate) const CAIRO_NATIVE_BINARY_NAME: &str = "starknet-native-compile";
```

**Review by ArniStarkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1998)* status: 2 of 10 files reviewed, 3 unresolved discussions (waiting on @avi-starkware and @Yoni-Starkware)

___
*[`crates/starknet_sierra_compile/src/command_line_compiler.rs` line 28 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/1998#-OBZolRN4mvlNkNUMiDe:-OBZwiQn3qtMVOVoR2XC:brbxs86) ([raw file](https://github.com/starkware-libs/sequencer/blob/6b16bfb4d9f32eef82023b09eaef0dd1855d59dd/crates/starknet_sierra_compile/src/command_line_compiler.rs#L28)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

Will this structure typically be used for the compilations `Sierra` -> `Casm` *and* `Sierra` -> `Cairo-native`?
Will there be a case where we want to have the second feature without the first? In that case, setting up the first feature is redundant. Maybe clever use of traits can save code duplication here.
</blockquote></details>

Talked F2F.
1. Once the build step is complete, struct and its implementation are simple. We save one more path and have one simple impl.
2. There might be uses for both compilations in the same scope.
3. We can postpone this decoupling, and advance the future (non-blocking).

**Review by ArniStarkware — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1998)* status: 2 of 10 files reviewed, 4 unresolved discussions (waiting on @avi-starkware and @Yoni-Starkware)

___
*[`crates/starknet_sierra_compile/src/command_line_compiler.rs` line 88 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/1998#-OBZyIv9-MYCAPKuFb4B:-OBZyIv9-MYCAPKuFb4C:b-b1dkb2) ([raw file](https://github.com/starkware-libs/sequencer/blob/6b16bfb4d9f32eef82023b09eaef0dd1855d59dd/crates/starknet_sierra_compile/src/command_line_compiler.rs#L88)):*
> ```XML
>             "Path to starknet native compile binary: {:?}",
>             self.path_to_starknet_native_compile_binary.as_os_str()
>         );
> ```

Delete the print  (I assume it serves no purpose).

**Review by avi-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1998)* status: 2 of 10 files reviewed, 1 unresolved discussion (waiting on @ArniStarkware and @Yoni-Starkware)

___
*[`crates/starknet_sierra_compile/src/build_utils.rs` line 5 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/1998#-OBZo-aT5-7zks4SDF5A:-OBZu4e_4r8dDYY2zTH-:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/6b16bfb4d9f32eef82023b09eaef0dd1855d59dd/crates/starknet_sierra_compile/src/build_utils.rs#L5)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

Or is it redundant?
</blockquote></details>

Done.
___
*[`crates/starknet_sierra_compile/src/command_line_compiler.rs` line 88 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/1998#-OBZyIv9-MYCAPKuFb4B:-OBaSpvxDKA1RTNBSg2m:b6rs0km) ([raw file](https://github.com/starkware-libs/sequencer/blob/6b16bfb4d9f32eef82023b09eaef0dd1855d59dd/crates/starknet_sierra_compile/src/command_line_compiler.rs#L88)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

Delete the print  (I assume it serves no purpose).
</blockquote></details>

Done.
Nice catch!

**Review by Yoni-Starkware — CHANGES_REQUESTED** [review]
> Reviewed 3 of 10 files at r1, 1 of 2 files at r2, 2 of 4 files at r3, 1 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1998)* status: 7 of 10 files reviewed, 4 unresolved discussions (waiting on @ArniStarkware, @avi-starkware, and @graphite-app[bot])

___
*[`crates/starknet_sierra_compile/src/command_line_compiler.rs` line 80 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/1998#-OBdYM7N1L5Tc2MnaJQA:-OBdYM7N1L5Tc2MnaJQB:b-nwyuxv) ([raw file](https://github.com/starkware-libs/sequencer/blob/14254323f1f39815ac76327b31fc542dce7d25d0/crates/starknet_sierra_compile/src/command_line_compiler.rs#L80)):*
> ```XML
> #[cfg(feature = "cairo_native")]
> impl SierraToNativeCompiler for CommandLineCompiler {
>     fn compile_to_native(
> ```

Add TODO to limit the process run by time and memory
___
*[`crates/starknet_sierra_compile/src/command_line_compiler.rs` line 106 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/1998#-OBdY4Yn96ghkSn-hDtC:-OBdY4Yn96ghkSn-hDtD:b-5w7l1q) ([raw file](https://github.com/starkware-libs/sequencer/blob/14254323f1f39815ac76327b31fc542dce7d25d0/crates/starknet_sierra_compile/src/command_line_compiler.rs#L106)):*
> ```XML
>             return Err(CompilationUtilError::CompilationError(stderr_output));
>         };
>         Ok(AotContractExecutor::load(Path::new(&output_file_path))?)
> ```

You can share most of this code with the existing compile func. Please do so.
Also, be consistent with the output destination



_Code quote:_
```XML
        // Create a temporary file to store the Sierra contract class.
        let serialized_contract_class = serde_json::to_string(&contract_class)?;

        let mut temp_file = NamedTempFile::new()?;
        temp_file.write_all(serialized_contract_class.as_bytes())?;
        let temp_file_path = temp_file.path().to_str().ok_or(
            CompilationUtilError::UnexpectedError("Failed to get temporary file path".to_owned()),
        )?;

        let output_file_path = output_file_path();

        // Set the parameters for the compile process.
        let mut command = Command::new(self.path_to_starknet_native_compile_binary.as_os_str());
        command.args([temp_file_path, &output_file_path]);

        let compile_output = command.output()?;

        if !compile_output.status.success() {
            let stderr_output = String::from_utf8(compile_output.stderr)
                .unwrap_or("Failed to get stderr output".into());
            return Err(CompilationUtilError::CompilationError(stderr_output));
        };
        Ok(AotContractExecutor::load(Path::new(&output_file_path))?)
```
___
*[`crates/starknet_sierra_compile/build.rs` line 16 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/1998#-OBdVjIzC5oa-8lhzvIw:-OBdVjJ-6jaustFDZQxT:b-naobih) ([raw file](https://github.com/starkware-libs/sequencer/blob/14254323f1f39815ac76327b31fc542dce7d25d0/crates/starknet_sierra_compile/build.rs#L16)):*
> ```XML
> const REQUIRED_CAIRO_LANG_VERSION: &str = "2.7.1";
> #[cfg(feature = "cairo_native")]
> const REQUIRED_CAIRO_NATIVE_VERSION: &str = "0.2.0-alpha.4";
> ```

Why not `0.2.1-alpha.0`?



_Code quote:_
```XML
 "0.2.0-alpha.4";
```

**Review by avi-starkware — COMMENTED** [review]
> Reviewed 1 of 4 files at r3, 1 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1998)* status: 9 of 10 files reviewed, 3 unresolved discussions (waiting on @Yoni-Starkware)

___
*[`crates/starknet_sierra_compile/build.rs` line 16 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/1998#-OBdVjIzC5oa-8lhzvIw:-OBdk3ov3_5VxSx35Hzc:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/14254323f1f39815ac76327b31fc542dce7d25d0/crates/starknet_sierra_compile/build.rs#L16)):*
<details><summary><i>Previously, Yoni-Starkware (Yoni) wrote…</i></summary><blockquote>

Why not `0.2.1-alpha.0`?
</blockquote></details>

Done.
___
*[`crates/starknet_sierra_compile/src/command_line_compiler.rs` line 80 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/1998#-OBdYM7N1L5Tc2MnaJQA:-OBdsOFC7A31Mx3GcbfY:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/14254323f1f39815ac76327b31fc542dce7d25d0/crates/starknet_sierra_compile/src/command_line_compiler.rs#L80)):*
<details><summary><i>Previously, Yoni-Starkware (Yoni) wrote…</i></summary><blockquote>

Add TODO to limit the process run by time and memory
</blockquote></details>

Done.
___
*[`crates/starknet_sierra_compile/src/command_line_compiler.rs` line 106 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/1998#-OBdY4Yn96ghkSn-hDtC:-OBdsPDh6zVEpgm5MYFD:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/14254323f1f39815ac76327b31fc542dce7d25d0/crates/starknet_sierra_compile/src/command_line_compiler.rs#L106)):*
<details><summary><i>Previously, Yoni-Starkware (Yoni) wrote…</i></summary><blockquote>

You can share most of this code with the existing compile func. Please do so.
Also, be consistent with the output destination
</blockquote></details>

Done.

**Comment by avi-starkware** [comment]
*[`crates/starknet_sierra_compile/src/command_line_compiler.rs` line 76 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/1998#-OBe6P5146TvmSjF1aIH:-OBeF9f671YL0mNZqaN4:b5vp72m) ([raw file](https://github.com/starkware-libs/sequencer/blob/c33a0942d473855ef28f87e44c74353c5d8dca97/crates/starknet_sierra_compile/src/command_line_compiler.rs#L76)):*
<details><summary><i>Previously, Yoni-Starkware (Yoni) wrote…</i></summary><blockquote>

Use stdout for consistency with sierra>casm
</blockquote></details>

We will need to save the binary to a file anyway, though. The executor loads the binary it executes from the path, not from a struct. Moreover, there is an additional `.json` file with the compiled binary (at the same path, same file name, with `.json extension), which contains the metadata required to construct the executor.

In summary, to use stdout the following step would have to happen:
1. The `AotContractExecutor` is saved to a binary and json file.
2. The binary and json files are read and printed to stdout.
3. Stdout is parsed to obtain two strings: one for the binary and one for the json.
4. Both strings are saved as files.
5. The `AotContractExecutor` is loaded from these two files and returned as a final result.

**Review by avi-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1998)* status: 6 of 11 files reviewed, 4 unresolved discussions (waiting on @ArniStarkware, @graphite-app[bot], and @Yoni-Starkware)

___
*[`crates/starknet_sierra_compile/src/command_line_compiler.rs` line 74 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/1998#-OBe6YzfAxl6NOWwZKhi:-OBeF8dJ0jiF5cphmjdc:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/c33a0942d473855ef28f87e44c74353c5d8dca97/crates/starknet_sierra_compile/src/command_line_compiler.rs#L74)):*
<details><summary><i>Previously, Yoni-Starkware (Yoni) wrote…</i></summary><blockquote>

Move to the shared func

Actually, you can share the entire code. Pass to the shared func `path_to_binary` and `additional_args`.
</blockquote></details>

Done.

**Review by Yoni-Starkware — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1998)* status: 5 of 11 files reviewed, 1 unresolved discussion (waiting on @ArniStarkware, @avi-starkware, and @graphite-app[bot])

___
*[`crates/starknet_sierra_compile/src/command_line_compiler.rs` line 76 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/1998#-OBe6P5146TvmSjF1aIH:-OBeOp2oAcR4R4in1TnY:b-llp3oz) ([raw file](https://github.com/starkware-libs/sequencer/blob/c33a0942d473855ef28f87e44c74353c5d8dca97/crates/starknet_sierra_compile/src/command_line_compiler.rs#L76)):*
<details><summary><i>Previously, avi-starkware wrote…</i></summary><blockquote>

We will need to save the binary to a file anyway, though. The executor loads the binary it executes from the path, not from a struct. Moreover, there is an additional `.json` file with the compiled binary (at the same path, same file name, with `.json extension), which contains the metadata required to construct the executor.

In summary, to use stdout the following step would have to happen:
1. The `AotContractExecutor` is saved to a binary and json file.
2. The binary and json files are read and printed to stdout.
3. Stdout is parsed to obtain two strings: one for the binary and one for the json.
4. Both strings are saved as files.
5. The `AotContractExecutor` is loaded from these two files and returned as a final result.
</blockquote></details>

Okay, no need. But what happens to the output file after you are done?
maybe you should use a temp file.

**Review by Yoni-Starkware — CHANGES_REQUESTED** [review]
> Reviewed 1 of 4 files at r6.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1998)* status: 6 of 11 files reviewed, 3 unresolved discussions (waiting on @ArniStarkware, @avi-starkware, and @graphite-app[bot])

___
*[`crates/starknet_sierra_compile/src/compile_test.rs` line 64 at r8](https://reviewable.io/reviews/starkware-libs/sequencer/1998#-OBePexhDVjD9zxnCVnb:-OBePexhDVjD9zxnCVnc:b-hquqyx) ([raw file](https://github.com/starkware-libs/sequencer/blob/6e8a5c96b8453ad371711ff4c7e0017699587a72/crates/starknet_sierra_compile/src/compile_test.rs#L64)):*
> ```XML
>     let compiler = command_line_compiler();
>     env::set_current_dir(get_absolute_path(TEST_FILES_FOLDER)).expect("Failed to set current dir.");
>     let sierra_path = Path::new(FAULTY_ACCOUNT_CLASS_FILE);
> ```

Share this code using a fixture



_Code quote:_
```XML
    env::set_current_dir(get_absolute_path(TEST_FILES_FOLDER)).expect("Failed to set current dir.");
    let sierra_path = Path::new(FAULTY_ACCOUNT_CLASS_FILE);
```
___
*[`crates/starknet_sierra_compile/src/errors.rs` line 4 at r7](https://reviewable.io/reviews/starkware-libs/sequencer/1998#-OBeP0Qq4hPLP97sdfbt:-OBeP0Qq4hPLP97sdfbu:b-iiffc) ([raw file](https://github.com/starkware-libs/sequencer/blob/6fb18cf0220f8d0f5f882d633829eeaa6c143716/crates/starknet_sierra_compile/src/errors.rs#L4)):*
> ```XML
> use cairo_lang_starknet_classes::casm_contract_class::StarknetSierraCompilationError;
> use thiserror::Error;
> ```

Restore the blank line.

**Comment by avi-starkware** [comment]
*[`crates/starknet_sierra_compile/src/errors.rs` line 4 at r7](https://reviewable.io/reviews/starkware-libs/sequencer/1998#-OBeP0Qq4hPLP97sdfbt:-OBeTHKT-HIONzErPp6C:b-u04y3r) ([raw file](https://github.com/starkware-libs/sequencer/blob/6fb18cf0220f8d0f5f882d633829eeaa6c143716/crates/starknet_sierra_compile/src/errors.rs#L4)):*
<details><summary><i>Previously, Yoni-Starkware (Yoni) wrote…</i></summary><blockquote>

Restore the blank line.
</blockquote></details>

If I do it fails rustfmt in the CI

**Review by Yoni-Starkware — COMMENTED** [review]
> Reviewed 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1998)* status: 8 of 11 files reviewed, 1 unresolved discussion (waiting on @ArniStarkware, @avi-starkware, and @graphite-app[bot])

**Review by avi-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1998)* status: 8 of 11 files reviewed, 1 unresolved discussion (waiting on @ArniStarkware, @graphite-app[bot], and @Yoni-Starkware)

___
*[`crates/starknet_sierra_compile/src/compile_test.rs` line 64 at r8](https://reviewable.io/reviews/starkware-libs/sequencer/1998#-OBePexhDVjD9zxnCVnb:-OBeZZof7Iezns_Y4rwJ:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/6e8a5c96b8453ad371711ff4c7e0017699587a72/crates/starknet_sierra_compile/src/compile_test.rs#L64)):*
<details><summary><i>Previously, Yoni-Starkware (Yoni) wrote…</i></summary><blockquote>

Share this code using a fixture
</blockquote></details>

Done.

**Review by Yoni-Starkware — COMMENTED** [review]
> Reviewed 1 of 4 files at r6.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1998)* status: 9 of 11 files reviewed, 1 unresolved discussion (waiting on @ArniStarkware and @graphite-app[bot])

**Review by Yoni-Starkware — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r10.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1998#-:-OBebVa3BHivMDfJa23e:bnfp4nl)* status: 10 of 11 files reviewed, all discussions resolved (waiting on @ArniStarkware and @graphite-app[bot])

**Review by Yoni-Starkware — COMMENTED** [review]
> Reviewed 2 of 2 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1998)* status: 10 of 11 files reviewed, all discussions resolved (waiting on @ArniStarkware and @graphite-app[bot])

**Review by Yoni-Starkware — COMMENTED** [review]
> Reviewed 2 of 2 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1998)* status: 10 of 11 files reviewed, all discussions resolved (waiting on @graphite-app[bot])

**Review by ArniStarkware — APPROVED** [review]
> Reviewed 1 of 2 files at r2, 1 of 4 files at r3, 2 of 4 files at r6, 1 of 2 files at r9, 1 of 2 files at r11, 2 of 2 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1998)* status: 10 of 11 files reviewed, all discussions resolved (waiting on @graphite-app[bot])

**Review by ArniStarkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1998)* status: 10 of 11 files reviewed, all discussions resolved (waiting on @graphite-app[bot] and @Yoni-Starkware)

___
*[`crates/starknet_sierra_compile/src/command_line_compiler.rs` line 52 at r12](https://reviewable.io/reviews/starkware-libs/sequencer/1998#-OBej7lGDnZNXpuZ0JZ-:-OBej7lGDnZNXpuZ0JZ0:bronm84) ([raw file](https://github.com/starkware-libs/sequencer/blob/83d4a11d26fd98b713aebd8eb6a71eaa5e55e5b3/crates/starknet_sierra_compile/src/command_line_compiler.rs#L52)):*
> ```XML
> 
>         let stdout = compile_with_args(compiler_binary_path, contract_class, &additional_args)?;
>         Ok(serde_json::from_slice::<CasmContractClass>(&stdout)?)
> ```

Non-blocking.

We should add an output dir parameter if starkness-sierra-compile supports it.
Not that I am sure how to do that.



_Suggestion:_
```XML
        let raw_casm_dir = ...;
        let additional_args = [
            "--add-pythonic-hints",
            "--max-bytecode-size",
            &self.config.max_bytecode_size.to_string(),
        ];

        let _stdout = compile_with_args(compiler_binary_path, contract_class, &additional_args)?;
        Ok(serde_json::from_slice::<CasmContractClass>(raw_casm_dir.into())?)
```

**Review by avi-starkware — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1998)* status: 10 of 11 files reviewed, all discussions resolved (waiting on @Yoni-Starkware)

**Review by avi-starkware — COMMENTED** [review]
> Reviewed 1 of 10 files at r1, 1 of 2 files at r2, 1 of 4 files at r3, 2 of 4 files at r6, 2 of 2 files at r9, 1 of 2 files at r11, 2 of 2 files at r12.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1998)* status: 10 of 11 files reviewed, all discussions resolved (waiting on @Yoni-Starkware)

**Review by avi-starkware — COMMENTED** [review]
> Reviewed 1 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1998)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @Yoni-Starkware)
